File: examples/java/src/main/java/com/xebia/functional/xef/java/auto/Weather.java
package com.xebia.functional.xef.java.auto;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

public class Weather {
    public List<String> answer;

    private static CompletableFuture<Void> clothesRecommend(AIScope scope) {
        return scope.prompt("Knowing this forecast, what clothes do you recommend I should wear?", Weather.class)
              .thenAccept(weather ->
                    System.out.println(weather.answer)
              );
    }

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        try (AIScope scope = new AIScope()) {
            scope.contextScope(scope.search("Weather in Cádiz, Spain").get(), Weather::clothesRecommend).get();
        }
    }

}

File: example/src/main/kotlin/com/xebia/functional/langchain4k/chain/Weather.kt
package com.xebia.functional.langchain4k.chain

import com.xebia.functional.auto.ai
import com.xebia.functional.auto.getOrElseSimilarity
import com.xebia.functional.tool.search
import io.github.oshai.KotlinLogging

suspend fun main() {
    val logger = KotlinLogging.logger("Weather")

    val question = "Knowing this forecast, what clothes do you recommend I should wear?"
    val answer: Map<String, String> = getQuestionAnswer(question)

    logger.info { answer }
}

private suspend fun getQuestionAnswer(
    question: String
): Map<String, String> = ai {
    agent(search("Weather in Cádiz, Spain")) {
        prompt<Map<String, String>>(question)
    }
}.getOrElseSimilarity {
    throw IllegalStateException(it.reason)
}
